### PR TITLE
introduce multimap

### PR DESCRIPTION
**Pull request by yoursunny**
MultiMap is a popular data structure that associates one key with several values.
This data structure is available in C++17 as [std::unordered\_multimap](https://en.cppreference.com/w/cpp/container/unordered_multimap).
[github.com/jwangsadinata/go-multimap](https://pkg.go.dev/github.com/jwangsadinata/go-multimap) has an untyped Go implementation of this data structure.

The API design in this code is borrowed from [NPM mnemonist/multi-map](https://yomguithereal.github.io/mnemonist/multi-map).
#6 should be merged before this PR.

### CODE REVIEW

**Comment by zyedidia** [comment]
Is there a significant difference between this and using a normal map/tree with an array or set as the value?

**Comment by yoursunny** [comment]
Compared to `map[K]map[V]struct{}`, the `MultiMap` offers a better API:

* During insertion, the caller does not need to check whether a sub-container (`map[V]struct{}`) exists.
* During deletion, the caller does not need to manually delete a sub-container when it becomes empty.
* There is a `Size()` method to determine total number of entries.
* There is an `Each()` method to iterate over all entries. The caller does not need to iterate outer and inner container separately.

File: multimap/multimap.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	// Count returns number of entries with a given key.
+	Count(key K) int
+	// Has determines whether at least one entry exists with a given key.
+	Has(key K) bool
+	// Get returns a list of values with a given key.
+	Get(key K) []V
+
+	// Set adds an entry.
+	// Whether duplicate entries are allowed depends on the chosen implementation.
+	Set(key K, value V)
```
Comment (zyedidia) [inline multimap/multimap.go:26]: Perhaps to be consistent with the other data structures this function should be called `Put`.

Comment (zyedidia) [inline multimap/multimap.go:26]: Perhaps to be consistent with the other data structures this function should be called `Put`.

**Comment by zyedidia** [comment]
It seems like there are multiple permutations for multimap implementation regarding whether to use distinct values and sorting for the keys/values. Would it be possible to use a more general Container interface (e.g., with get, put, each) and have the user provide the underlying implementation (by passing an AVL tree, set, slice, or a custom data structure)?

**Comment by zyedidia** [comment]
It seems like there are multiple permutations for multimap implementation regarding whether to use distinct values and sorting for the keys/values. Would it be possible to use a more general Container interface (e.g., with get, put, each) and have the user provide the underlying implementation (by passing an AVL tree, set, slice, or a custom data structure)?

**Comment by yoursunny** [comment]
> Would it be possible to use a more general Container interface (e.g., with get, put, each) and have the user provide the underlying implementation (by passing an AVL tree, set, slice, or a custom data structure)?

I believe it's infeasible to have one type for everything.

Take `NewMapSlice` vs `NewMapSet` as an example:

* `NewMapSlice` has `[V comparable]` type constraint.
* `NewMapSet` has `[V any]` type constraint.

If I create a general structure that accepts both, there's no way to declare `V` type constraint, because:

* Type constraint cannot be used as type assertion at runtime.
* Go 1.18 does not support specialization in generics.

Instead of `NewMapSet`, it is feasible to have:

```go
func NewComparableKeyAnyValue[K comparable, V any, VC ValueContainer<V>](implK comparableKeyImpl[K, VC], implV VC, valueLess g.LessFn[V]) MultiMap[K, V]
```

But then you still end up with a separate function for each combination of comparable vs any.
The only gained ability is replacing `mapMultiMap` and `valuesSet` with something else.
However, the constructor becomes much more complicated.

**Comment by zyedidia** [comment]
Ok I see the problem, thanks. I think it might be more consistent to also rename `Delete` to `RemoveAll` or `RemoveAssociation` (which is also more consistent with `EachAssociation`) -- what do you think? Otherwise I think this is ready to merge.

**Comment by yoursunny** [comment]
It's `RemoveAll` now.

**Pull request merged**